### ISSUE TITLE
vectorAddMMAP target does not compile due to virtual memory feature set_access_mode

### ISSUE THREAD

**Issue by Q-Minh** [eyalroz/cuda-api-wrappers#344]
The example target `vectorAddMMAP` has the error `cuda-api-wrappers\examples\modified_cuda_samples\vectorAddMMAP\vectorAddMMAP.cpp(163,107): error C2665: 'cuda::memory::virtual_::set_access_mode': none of the 2 overloads could convert all the argument types` when compiling with MSVC using VS2019.

**Comment by eyalroz** [comment]
Why couldn't they? What are the types MSVC is seeing? :-(

**Comment by Q-Minh** [comment]
There is an ambiguous call to either:
```
template <template <typename... Ts> class ContiguousContainer>
inline void set_access_mode(
	mapping_t mapping,
	const ContiguousContainer<device_t>& devices,
	access_mode_t access_mode);
``` 
or 
```
template <template <typename... Ts> class ContiguousContainer>
void set_access_mode(
	region_t fully_mapped_region,
	const ContiguousContainer<device_t>& devices,
	access_mode_t access_mode);
```
in `vectorAddMMAP.cpp` line 163. This is quite weird, to be honest, looks like a compiler bug to me, since the signatures are quite obviously different, with `mapping_t` and `region_t` having no related inheritance tree or anything.

**Comment by eyalroz** [comment]
Maybe it's an issue with MSVC's support for template-template parameters.

What happens if we separate the two function names? i.e. `set_mapping_access_mode()` and `set_region_access_mode()`?

**Comment by Q-Minh** [comment]
Okay, so the actual problem was that `std::iterator_traits<Iter>` was not able to specialize `Iter = enumerator<...>::const_iterator`, because there were some typedefs/aliases missing. The fix is then to add:
```
using difference_type = std::ptrdiff_t;
using value_type = value_type;
using pointer = value_type*;
using reference = value_type&;
using iterator_category = std::forward_iterator_tag;
```
in the declaration of struct `enumerator<Container>::iterator`, and
```
using difference_type = std::ptrdiff_t;
using value_type = value_type;
using pointer = value_type const*;
using reference = value_type const&;
using iterator_category = std::forward_iterator_tag;
```
in the declaration of struct `enumerator<Container>::const_iterator`.

Now MSVC can disambiguate between the 2 overloads of `set_access_mode(...)`.

**Comment by eyalroz** [comment]
Are you sure we need value_type? That looks redundant. Also, are you sure we need to add all of these into the iterator, rather than some type definitions  to the enumerator class (e.g. pointer, const_pointer etc.)?

**Comment by Q-Minh** [comment]
Yeah, pretty sure, because [`std::iterator_traits<Iter>`](https://en.cppreference.com/w/cpp/iterator/iterator_traits) takes the iterator itself as template parameter, not the container, and it attempts to directly access member typedefs on the `Iter` type passed in.

**Comment by eyalroz** [comment]
How's this?

**Comment by Q-Minh** [comment]
> How's this?

Looks good! I don't know about gcc and clang, but at least with msvc, you don't need to specify `enumerator::` to capture the `value_type` of the enclosing `enumerator` class. However, if you do want to specify it, I would write `enumerator<Container>::value_type` instead.

**Comment by eyalroz** [comment]
@Q-Minh : Sure.
